builder/builder.go

```diff
@@ -426,93 +442,205 @@ func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+			if err != nil {
+				return nil, err
+			} else if len(includes) > 0 {
+				bundleSet, err = b.buildMixBundleSetFull(includes, bundleSet)
+				if err != nil {
+					return nil, err
 				}
 			}
+		}
+	}
+	return bundleSet, nil
+}
 
-			fmt.Printf("Adding bundle %q\n", bundle)
-			if err = helpers.CopyFile(filepath.Join(b.Bundledir, bundle), filepath.Join(clrbundledir, bundle)); err != nil {
-				return bundleAddCount, err
-			}
-			bundleAddCount++
+// getMixBundlesListFull returns a sorted list of all bundles needed for the
+// mix, including all bundles in the Mix Bundles List and any those include.
+func (b *Builder) getMixBundlesListFull() ([]string, error) {
+	bundles, err := b.getMixBundlesList()
+	if err != nil {
+		return nil, err
+	}
+
+	bundleSet := stringset.New()
+	bundleSet, err = b.buildMixBundleSetFull(bundles, bundleSet)
+	if err != nil {
+		return nil, err
+	}
+
+	return bundleSet.Sort(), nil
+}
+
+// AddBundles adds the specified bundles to the Mix Bundles List. Values are
+// verified as valid, and duplicate values are removed. The resulting Mix
+// Bundles List will be in sorted order.
+//
+// bundles: array slice of bundle names
+// all: include all CLR bundles. Overrides bundles.
+// git: automatically git commit with bundles added
+func (b *Builder) AddBundles(bundles []string, alllocal bool, allupstream bool, git bool) error {
+	// Fetch upstream bundles if needed
+	if err := b.getUpstreamBundles(b.Clearver, true); err != nil {
+		return err
+	}
+
+	// Read in current mix bundles list
+	bundleSet, err := b.readMixBundlesListAsSet()
+	if err != nil {
+		return err
+	}
+
+	// Check that the ones passed in exist
+	for _, bundle := range bundles {
+		path, err := b.getBundlePath(bundle)
+		if err != nil {
+			return err
+		}
+		if path == filepath.Join(b.Lbundledir, bundle) {
+			fmt.Printf("Adding bundle %s from local bundles\n", bundle)
 		} else {
-			fmt.Printf("Warning: bundle %q already exists; skipping.\n", bundle)
+			fmt.Printf("Adding bundle %s from upstream bundles\n", bundle)
 		}
 	}
-	// Recurse on included bundles
-	if len(includes) > 0 {
-		ib, err := b.AddBundles(includes, force, false, false)
+
+	// Add all local bundles to the bundles
+	if alllocal {
+		files, err := ioutil.ReadDir(b.Lbundledir)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to read local bundles dir: %s", b.Lbundledir)
+		}
+
+		for _, file := range files {
+			bundle := file.Name()
+			bundles = append(bundles, bundle)
+			fmt.Printf("Adding bundle %s from local bundles\n", bundle)
+		}
+	}
+
+	// Add all upstream bundles to the bundles
+	if allupstream {
+		clrbundledir := b.getUpstreamBundlesPath(b.Clearver)
+		files, err := ioutil.ReadDir(clrbundledir)
 		if err != nil {
-			return bundleAddCount, err
+			return errors.Wrapf(err, "Failed to read upstream bundles dir: %s", clrbundledir)
+		}
+
+		for _, file := range files {
+			bundle := file.Name()
+			bundles = append(bundles, bundle)
```

Comment:
why not just 
`bundles = append(bundles, file.Name())` instead of creating a new var each time?